doc/architecture/adr-001.md

```diff
@@ -0,0 +1,117 @@
+# ADR 001: Add CosmWasm Module
+
+## Changelog
+* 05-03-2021: Initial Draft
+
+## Context
+
+Crypto.org Chain logic may need to be extended to support new digital asset and financial use cases
+in the near future. 
+In the initial implementation, Crypto.org Chain logic is extensible through three different mechanisms:
+
+1. Cryptographic commitments and multi-signatures: 
+through the existing and future (e.g. BIP-340) signature verification mechanisms, 
+one can potentially create accounts that, through their construction, enforce additional conditions
+on their transactions. This could done in different ways (e.g. see "Discreet Log Contracts" and "Adaptor Signatures").
+Alternatively, the most "flexible" approach is if a conventional multi-signature is constructed, but one of the
+"signers" is a secure enclave program executed by all parties.
+
+2. Inter-Blockchain Communication (IBC): the other connecting network or solo machine could enforce additional conditions
+or provide extra functionality (over the current ICS20).
+
+3. The Cosmos SDK "upgrade" module: the network-wide activation of new modules or changes in existing modules can be coordinated
+via the dedicated module.
+
+While the first two mechanisms are intriguing, their current ecosystem support is limited or non-existent.
+Besides that, the extra logic they can provide is mostly opaque to the overall network: this is often desirable
+(for scalability reasons), but may not desirable in certain use cases where the full public verifiability is needed.
+
+The third mechanism is a bit "heavy-weight" in the sense that it requires an explicit action by all node operators
+at the same time. The current version of Tendermint (0.34) also does not allow a "soft" upgrade.
+
+CosmWasm is a new promising module for Cosmos SDK-based projects. It allows extensibility through the 
+dominant approach of metered "stored procedures", popularized by Ethereum under the term "smart contracts". 
+Unlike on Ethereum, the target binary format (WebAssembly) is an official LLVM target 
+and mainstream programming languages (such as Rust) can be used for development.
+The overall design avoids many of the common pitfalls (e.g. reetrancy attacks) developers face on Ethereum.
+
+There are two main concerns regarding CosmWasm:
+
+1. It is a new unstable module: this is concerning from both the security perspective
+(e.g. DoS attack vectors in its execution engine) and the backwards compatibility perspective
+(stored bytecode that targeted the older module version may not execute correctly on its new version).
```

Comment:
There have been a number of breaking CosmWasm changes in the byte-code to host interface. After CosmWasm 0.11, we kept that interface stable for a few versions while doing major improvements to the tooling (0.11, 0.12, 0.13). All nodes must run the same version (eg. different gas costs) but the same wasm code would work on all of them.

We are preparing a 0.14 release with quite a few breaking changes that we have been waiting for for some time. 

After this, we aim to get feedback and maybe one more pass of API changes before stabilizing a 1.0 CosmWasm. Once we commit to CosmWasm 1.0, the wasm interface will be fixed and you will not have to worry about such breaking changes on upgrades.